[CSS Zoom] Apply zoom factor to text-shadow, box-shadow, and drop-shadow

brentfulgham opened the pull request:
#### a797d90e51ab84dde8d6cb343ac13ee43df735d2
<pre>
[CSS Zoom] Apply zoom factor to text-shadow, box-shadow, and drop-shadow
<a href="https://bugs.webkit.org/show_bug.cgi?id=300577">https://bugs.webkit.org/show_bug.cgi?id=300577</a>
<a href="https://rdar.apple.com/162463179">rdar://162463179</a>

Reviewed by Tim Nguyen.

Update TextShadow, BoxShadow, and DropShadow to use the new strong type system,
and apply zoom factor at evaluation-time.

This change consists of changing Length&lt;&gt; to Length&lt;CSS:UnzoomedAll&gt; in shadow
related data, Length&lt;Nonnegative&gt; to Length&lt;CSS::NonnegativeUnzoomed&gt; in blur
related data, and threading the Style zoom state through more of the callstack.

Test: imported/w3c/web-platform-tests/css/css-viewport/zoom/box-shadow.html
      imported/w3c/web-platform-tests/css/css-viewport/zoom/filters-drop-shadow.html
      imported/w3c/web-platform-tests/css/css-viewport/zoom/svg-computed-style.html
      imported/w3c/web-platform-tests/css/css-viewport/zoom/text-shadow.html

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/box-shadow-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/box-shadow.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/filters-drop-shadow-expected.html
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/filters-drop-shadow.html
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/reference/box-shadow-ref.html
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/reference/filters-drop-shadow-ref.html
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/svg-computed-style-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/svg-computed-style.html:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Background.cpp:
(WebCore::CSSPropertyParserHelpers::consumeSingleUnresolvedBoxShadow):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Filter.cpp:
(WebCore::CSSPropertyParserHelpers::consumeFilterDropShadow):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+TextDecoration.cpp:
(WebCore::CSSPropertyParserHelpers::consumeSingleUnresolvedTextShadow):
* Source/WebCore/css/values/borders/CSSBoxShadow.h:
* Source/WebCore/css/values/filter-effects/CSSDropShadowFunction.h:
* Source/WebCore/css/values/text-decoration/CSSTextShadow.h:
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::fontAttributesAtSelectionStart):
* Source/WebCore/editing/FontAttributeChanges.cpp:
(WebCore::cssValueForTextShadow):
* Source/WebCore/layout/formattingContexts/inline/display/InlineDisplayContentBuilder.cpp:
(WebCore::Layout::computeInkOverflowForInlineLevelBox):
(WebCore::Layout::InlineDisplayContentBuilder::appendTextDisplayBox):
* Source/WebCore/rendering/BackgroundPainter.cpp:
(WebCore::applyBoxShadowForBackground):
(WebCore::BackgroundPainter::paintBoxShadow const):
* Source/WebCore/rendering/EllipsisBoxPainter.cpp:
(WebCore::EllipsisBoxPainter::paint):
* Source/WebCore/rendering/LegacyInlineFlowBox.cpp:
(WebCore::LegacyInlineFlowBox::addTextBoxVisualOverflow):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::applyVisualEffectOverflow const):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::repaintAfterLayoutIfNeeded):
* Source/WebCore/rendering/TextBoxPainter.h:
(WebCore::TextBoxPainter::rotateShadowOffset):
* Source/WebCore/rendering/TextDecorationPainter.cpp:
(WebCore::TextDecorationPainter::paintBackgroundDecorations):
* Source/WebCore/rendering/TextPainter.cpp:
* Source/WebCore/rendering/svg/RenderSVGBlock.cpp:
(WebCore::RenderSVGBlock::computeOverflow):
* Source/WebCore/rendering/svg/RenderSVGText.cpp:
(WebCore::RenderSVGText::repaintRectInLocalCoordinates const):
(WebCore::RenderSVGText::updatePositionAndOverflow):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGModelObject.cpp:
(WebCore::adjustRectForOutlineAndShadow):
(WebCore::LegacyRenderSVGModelObject::outlineBoundsForRepaint const):
* Source/WebCore/style/values/borders/StyleBoxShadow.h:
(WebCore::Style::paintingSpread):
* Source/WebCore/style/values/borders/StyleShadow.h:
(WebCore::Style::requires):
(WebCore::Style::paintingExtent):
(WebCore::Style::paintingExtentAndSpread):
(WebCore::Style::shadowOutsetExtent):
(WebCore::Style::shadowInsetExtent):
(WebCore::Style::shadowHorizontalExtent):
(WebCore::Style::shadowVerticalExtent):
(WebCore::Style::shadowBlockDirectionExtent):
(WebCore::Style::shadowInlineDirectionExtent):
(WebCore::Style::adjustRectForShadow):
* Source/WebCore/style/values/filter-effects/StyleDropShadowFunction.cpp:
(WebCore::Style::toCSSDropShadow):
(WebCore::Style::createFilterOperation):
* Source/WebCore/style/values/text-decoration/StyleTextShadow.cpp:
(WebCore::Style::ToStyle&lt;CSS::TextShadow&gt;::operator):
* Source/WebCore/style/values/text-decoration/StyleTextShadow.h:
(WebCore::Style::paintingSpread):

Canonical link: <a href="https://commits.webkit.org/301596@main">https://commits.webkit.org/301596@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4f91a19d67b0a31e56f78472f8007734eea92e0c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/126458 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/46104 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/37010 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/133361 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/78148 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/8b02ef45-a222-40ec-8444-06fca5959457) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/128329 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46748 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54644 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/96241 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/64331 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/e429c0b6-8771-43d5-af2e-4bc07b729cf2) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/129406 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37391 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/113095 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/76714 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/640b081e-d23f-417a-92b3-0945d90be394) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/36283 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/31272 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/76667 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/107198 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/31547 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/135898 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/53156 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40866 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/104744 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53639 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/109408 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/104445 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26642 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49914 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/28243 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/50560 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/53074 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/58891 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/52364 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55698 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/54095 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Background.cpp:
```diff
@@ -355,8 +355,8 @@ static std::optional<CSS::BoxShadow> consumeSingleUnresolvedBoxShadow(CSSParserT
+    std::optional<CSS::Length<CSS::AllUnzoomed>> x;
+    std::optional<CSS::Length<CSS::AllUnzoomed>> y;
     std::optional<CSS::Length<CSS::Nonnegative>> blur;
     std::optional<CSS::Length<>> spread;
```
Unfortunately not covered by the test, but blur and spread should also take in the zoom factor: https://drafts.csswg.org/css-viewport/#zoom-property

all `<length>` values should be covered, except those computing to auto or a percentage

nt1m replied:
Here and through out the PR, we should also apply zoom to those. The spec says to zoom all `<length>` values.

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Filter.cpp:
```diff
@@ -137,8 +137,8 @@ static std::optional<CSS::DropShadowFunction> consumeFilterDropShadow(CSSParserT
-    std::optional<CSS::Length<>> y;
+    std::optional<CSS::Length<CSS::AllUnzoomed>> x;
+    std::optional<CSS::Length<CSS::AllUnzoomed>> y;
     std::optional<CSS::Length<CSS::Nonnegative>> stdDeviation;
```
ditto here

nt1m reviewed (commented):
* This also changes box-shadow and `filter: drop-shadow()` (which is correct) but we should add a reference test for both of them. I think we can do both in one patch since the two types are intertwined.
* You can add an entry in css/css-viewport/zoom/svg-computed-style.html to test the SVG computed style:

```
    "text-shadow": {
      "value": "1px 2px 3px rgb(255, 255, 255)",
      "otherValues": ["2px 3px rgb(0, 255, 255)", "4px 1px rgb(0, 0, 255)"]
    },
```

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+TextDecoration.cpp:
```diff
@@ -55,8 +55,8 @@ static std::optional<CSS::TextShadow> consumeSingleUnresolvedTextShadow(CSSParse
-    std::optional<CSS::Length<>> y;
+    std::optional<CSS::Length<CSS::AllUnzoomed>> x;
+    std::optional<CSS::Length<CSS::AllUnzoomed>> y;
     std::optional<CSS::Length<CSS::Nonnegative>> blur;
```
ditto for this length, this should also use the new zoom architecture

nullhook commented on Source/WebCore/style/values/text-decoration/StyleTextShadow.h:
```diff
@@ -94,7 +95,7 @@ constexpr bool isInset(const TextShadow&)
 }
 
-constexpr LayoutUnit paintingSpread(const TextShadow&)
+constexpr LayoutUnit paintingSpread(const TextShadow&, const Style::ZoomFactor&)
```
unused ZoomFactor param?

nt1m replied:
> unused ZoomFactor param?

Yeah, text-shadows don't have a spread parameter like box-shadows

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/box-shadow.html:
```diff
@@ -0,0 +1,19 @@
+<title>CSS zoom applies to box-shadow when specified and inherited</title>
+<link rel="author" title="Brent Fulgham" href="mailto:bfulgham@apple.com">
+<link rel="help" href="https://drafts.csswg.org/css-viewport/">
+<link rel="match" href="reference/text-shadow-ref.html">
```
```suggestion
<link rel="match" href="reference/box-shadow-ref.html">
```

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/box-shadow-expected.html:
We need a copy of this in `css/css-viewport/zoom/reference/box-shadow-ref.html` for this to run in the WPT runner

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Filter.cpp:
We also need a test for `filter: drop-shadow(rgb(0,0,0) 2px 2px 2px)` similar to the `box-shadow` one, ditto in the svg-computed-style.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/svg-computed-style.html:
```diff
@@ -26,7 +26,15 @@
+    "box-shadow": {
+      "value": "rgb(0, 0, 0) 5px 5px 5px 0px",
+      "otherValues": ["rgb(0, 0, 0) 10px 10px 10px 0px"]
+    },
```
nit: let's keep this test in alphabetical ordering

brentfulgham committed:
[CSS Zoom] Apply zoom factor to text-shadow, box-shadow, and drop-shadow
https://bugs.webkit.org/show_bug.cgi?id=300577
rdar://162463179

Reviewed by Tim Nguyen.

Update TextShadow, BoxShadow, and DropShadow to use the new strong type system,
and apply zoom factor at evaluation-time.

This change consists of changing Length<> to Length<CSS:UnzoomedAll> in shadow
related data, Length<Nonnegative> to Length<CSS::NonnegativeUnzoomed> in blur
related data, and threading the Style zoom state through more of the callstack.

Test: imported/w3c/web-platform-tests/css/css-viewport/zoom/box-shadow.html
      imported/w3c/web-platform-tests/css/css-viewport/zoom/filters-drop-shadow.html
      imported/w3c/web-platform-tests/css/css-viewport/zoom/svg-computed-style.html
      imported/w3c/web-platform-tests/css/css-viewport/zoom/text-shadow.html

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/box-shadow-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/box-shadow.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/filters-drop-shadow-expected.html
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/filters-drop-shadow.html
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/reference/box-shadow-ref.html
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/reference/filters-drop-shadow-ref.html
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/svg-computed-style-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/svg-computed-style.html:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Background.cpp:
(WebCore::CSSPropertyParserHelpers::consumeSingleUnresolvedBoxShadow):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Filter.cpp:
(WebCore::CSSPropertyParserHelpers::consumeFilterDropShadow):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+TextDecoration.cpp:
(WebCore::CSSPropertyParserHelpers::consumeSingleUnresolvedTextShadow):
* Source/WebCore/css/values/borders/CSSBoxShadow.h:
* Source/WebCore/css/values/filter-effects/CSSDropShadowFunction.h:
* Source/WebCore/css/values/text-decoration/CSSTextShadow.h:
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::fontAttributesAtSelectionStart):
* Source/WebCore/editing/FontAttributeChanges.cpp:
(WebCore::cssValueForTextShadow):
* Source/WebCore/layout/formattingContexts/inline/display/InlineDisplayContentBuilder.cpp:
(WebCore::Layout::computeInkOverflowForInlineLevelBox):
(WebCore::Layout::InlineDisplayContentBuilder::appendTextDisplayBox):
* Source/WebCore/rendering/BackgroundPainter.cpp:
(WebCore::applyBoxShadowForBackground):
(WebCore::BackgroundPainter::paintBoxShadow const):
* Source/WebCore/rendering/EllipsisBoxPainter.cpp:
(WebCore::EllipsisBoxPainter::paint):
* Source/WebCore/rendering/LegacyInlineFlowBox.cpp:
(WebCore::LegacyInlineFlowBox::addTextBoxVisualOverflow):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::applyVisualEffectOverflow const):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::repaintAfterLayoutIfNeeded):
* Source/WebCore/rendering/TextBoxPainter.h:
(WebCore::TextBoxPainter::rotateShadowOffset):
* Source/WebCore/rendering/TextDecorationPainter.cpp:
(WebCore::TextDecorationPainter::paintBackgroundDecorations):
* Source/WebCore/rendering/TextPainter.cpp:
* Source/WebCore/rendering/svg/RenderSVGBlock.cpp:
(WebCore::RenderSVGBlock::computeOverflow):
* Source/WebCore/rendering/svg/RenderSVGText.cpp:
(WebCore::RenderSVGText::repaintRectInLocalCoordinates const):
(WebCore::RenderSVGText::updatePositionAndOverflow):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGModelObject.cpp:
(WebCore::adjustRectForOutlineAndShadow):
(WebCore::LegacyRenderSVGModelObject::outlineBoundsForRepaint const):
* Source/WebCore/style/values/borders/StyleBoxShadow.h:
(WebCore::Style::paintingSpread):
* Source/WebCore/style/values/borders/StyleShadow.h:
(WebCore::Style::requires):
(WebCore::Style::paintingExtent):
(WebCore::Style::paintingExtentAndSpread):
(WebCore::Style::shadowOutsetExtent):
(WebCore::Style::shadowInsetExtent):
(WebCore::Style::shadowHorizontalExtent):
(WebCore::Style::shadowVerticalExtent):
(WebCore::Style::shadowBlockDirectionExtent):
(WebCore::Style::shadowInlineDirectionExtent):
(WebCore::Style::adjustRectForShadow):
* Source/WebCore/style/values/filter-effects/StyleDropShadowFunction.cpp:
(WebCore::Style::toCSSDropShadow):
(WebCore::Style::createFilterOperation):
* Source/WebCore/style/values/text-decoration/StyleTextShadow.cpp:
(WebCore::Style::ToStyle<CSS::TextShadow>::operator):
* Source/WebCore/style/values/text-decoration/StyleTextShadow.h:
(WebCore::Style::paintingSpread):

Canonical link: https://commits.webkit.org/301596@main
